spring-web/src/main/java/org/springframework/http/HttpHeaders.java

```diff
@@ -927,8 +927,17 @@ public void setContentType(@Nullable MediaType mediaType) {
 	 */
 	@Nullable
 	public MediaType getContentType() {
-		String value = getFirst(CONTENT_TYPE);
-		return (StringUtils.hasLength(value) ? MediaType.parseMediaType(value) : null);
+		if (this.readOnly && this.cachedContentType != null) {
+			return this.cachedContentType;
+		}
+		else {
+			String value = getFirst(CONTENT_TYPE);
+			MediaType contentType = (StringUtils.hasLength(value) ? MediaType.parseMediaType(value) : null);
+			if (this.readOnly) {
+				this.cachedContentType = contentType;
+			}
+			return contentType;
+		}
```

Comment:
This can be improved. The `this.readOnly` check shows that this method is mixing two concerns: map lookup and caching. I would prefer if caching was done in a separate wrapper class(`ReadOnlyHttpHeaders`), that is returned when `HttpHeaders.readOnlyHttpHeaders` is called (instead of backing the map with an unmodifiable `MultiValueMap`).